core/src/main/java/com/datastax/oss/driver/internal/core/data/DefaultUdtValue.java

```diff
@@ -91,6 +92,39 @@ public ProtocolVersion protocolVersion() {
     return type.getAttachmentPoint().protocolVersion();
   }
 
+  @Override
```

Comment:
It might be better to use IntelliJ generated `equals` and `hashCode` methods which are more straight forward and depends on `ByteBuffer.equals()`:

```java
  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    DefaultUdtValue that = (DefaultUdtValue) o;
    return Objects.equals(type, that.type) && Arrays.equals(values, that.values);
  }

  @Override
  public int hashCode() {

    int result = Objects.hash(type);
    result = 31 * result + Arrays.hashCode(values);
    return result;
  }
```

Reply:
👍  for doing an instance equality check first (`this == o`).

About comparing the byte buffers directly, there is a comment in the 3.x implementation that explains why we might want to avoid that:

> Deserializing each value is slightly inefficient, but comparing
> the bytes could in theory be wrong (for varint for instance, 2 values
> can have different binary representation but be the same value due to
> leading zeros). So we don't take any risk.

Finally, rather than comparing `getClass()`, we should do an `instanceof` check, but against the interface, not the implementation. That way if someone writes their own `UdtValue` implementation, ours compares to it nicely. See `DefaultUserDefinedType.equals` for example.